Property docs (#6104)
microsoft/FluidFramework: docs/content/start/property-dds/property-schemas.md

---
title: Schemas
menuPosition: 2
---

A *Property Sets Schema* defines the structure of a Property Set. It defines which child properties exist in a
property and which types they have.

> The following example shows a possible schema for your rectangle. This example is only meant to show the various possibilities in defining schemas and is not necessarily showing the best way to define a rectangle.

Reviewer: ```suggestion
The following example shows a possible schema for your rectangle. This example is only meant to show the various possibilities in defining schemas and is not necessarily the best way to define a rectangle.
```

  "properties": [
    { "id": "position",
      "properties": [
        { "id": "x", "typeid": "Float64" },
        { "id": "y", "typeid": "Float64" }
      ]
    },
    { "id": "size", "typeid": "Float64", "context": "array", "length": 2 },
    { "id": "fillcolor", "typeid": "Sample:Color-1.0.0" },
    { "id": "border",
      "properties": [
        { "id": "color", "typeid": "Sample:Color-1.0.0" }
        { "id": "style", "typeid": "Enum",
          "properties": [
            { "id": "solid", "value": 1 },
            { "id": "dotted", "value": 2 },
            { "id": "dashed", "value": 3 }
          ]
        }
      ]
    },
    { "id": "caption", "typeid": "String" }
  ]
}
```

Here, we've defined size as an array of 2 elements (instead of using “width” and “height” properties) to demonstrate the use of the context, array, and length keywords.

```json
{ "id": "size", "typeid": "Float64", "context": "array", "length": 2 }
```

Also, the example shows that you can define nested complex types inline, as we did for position:

```json
{ "id": "position",
  "properties": [
    { "id": "x", "typeid": "Float64" },
    { "id": "y", "typeid": "Float64" }
  ]
}
```

> Or you can refer to an external schema, as we did for fillcolor.

Reviewer: ```suggestion
Or you can refer to an external schema, as we did for fill color.
```

      ...

      // Properties at a lower level
      { "id": "border",
        "properties": [
          { "id": "style", "typeid": "Enum",
            "properties": [
              // some properties here
            ]
          }
        ]
      },
      // other properties here
    ]
  }
```

## Sub-Property Field Name

All the entries in the properties array must be objects, and each must contain an id. This is the string that will later be used to retrieve the property to set or get its value.

## context

This identifier describes the type of collection of values the property contains. If not specified, it is assumed to be ‘single’.

Context types:

**single**
  <default> The sub-property contains a single instance of the given type.
**array**
  The sub-property contains an array of instances of the given type.
**map**
  The sub-property contains a map from strings to instances of the given type.
**set**
  The sub-property contains an unordered collection of NamedProperty.

## length

This is an optional keyword that can be used when context is ‘array’ to specify a fixed-length array. If no length is specified, the array length is dynamic.

> ## nnotation

Reviewer: ```suggestion
## Annotation
```

    { "id": "array", "typeid": "Int32", "value": [333], "context": "array" },
    { "id": "map", "typeid": "Int32", "context": "map", "value": { "key1": 333 } }
  ]
}
```

Example of polymorphic default values

```json
{
  "typeid": "Sample:Shape-1.0.0",
  "properties": [ { "id": "x", "typeid": "Float32", "value": 0 } ]
}

{
  "typeid": "Sample:Square-1.0.0",
  "inherits": "Sample:Shape-1.0.0",
  "properties": [ { "id": "side", "typeid": "Float32", "value": 0 } ]
}

{
  "typeid": "Sample:Space-1.0.0",
  "properties": [
    {
      "id": "arr", "typeid": "Sample:Shape-1.0.0", "context": "array",
      "typedValue": [
        { "typeid": "Sample:Square-1.0.0", "value": { "x": 110, "side": 110 } },
        { "typeid": "Sample:Shape-1.0.0", "value": { "x": 120} }
      ]
    }]
}
```

## Constants

A schema can define read-only properties through Constants. They are declared in a “constants” array which is at the same level as the properties one.

    Each constant should have an id, value and a valid typeid. The id must be unique per declared constants and properties, that is a constant and a property cannot have the same id.

>     A constant can be of primitive, enum, string, array, set, map, or custom type.

Reviewer: ```suggestion
A constant can be of primitive, enum, string, array, set, map, or custom type.
```

>     They are read only properties; values and instances can be retrieved like any other properties.

Reviewer: ```suggestion
They are read-only properties; values and instances can be retrieved like any other properties.
```

>     Constants can be overridden in inherited schemas. To do that, the inherited constant should keep the same 'id', 'typeid' and 'context' as the base one it overrides.

Reviewer: ```suggestion
Constants can be overridden in inherited schemas. To do that, the inherited constant should keep the same 'id', 'typeid' and 'context' as the base one it overrides.
```
